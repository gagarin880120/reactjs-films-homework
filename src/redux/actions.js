function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
