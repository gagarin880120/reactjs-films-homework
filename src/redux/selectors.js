import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => state.searchResults.map((v) => ({
  ...v,
  genres: getGenresString(v.genre_ids, state.genres),
}));

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;

export { searchResultsSelector, modalSelector, trailerSelector };
