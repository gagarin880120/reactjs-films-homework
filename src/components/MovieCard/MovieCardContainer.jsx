import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieCard from './MovieCard';
import { setModal, getTrailer } from '../../redux/actions';

export function MovieCardContainer({ onTrailerButtonClick, movie }) {
  return (
    <MovieCard
      onTrailerButtonClick={onTrailerButtonClick}
      movie={movie}
    />
  );
}

export const mapDispatchToProps = (dispatch) => ({
  onTrailerButtonClick(id) {
    dispatch(setModal(true));
    dispatch(getTrailer(id));
  },
});

MovieCardContainer.propTypes = {
  onTrailerButtonClick: PropTypes.func,
  movie: PropTypes.shape({}),
};

MovieCardContainer.defaultProps = {
  onTrailerButtonClick: null,
  movie: {},
};

export default connect(null, mapDispatchToProps)(MovieCardContainer);
