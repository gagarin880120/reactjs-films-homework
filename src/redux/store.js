import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
