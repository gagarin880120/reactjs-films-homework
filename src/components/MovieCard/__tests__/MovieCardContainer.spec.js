import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { MovieCardContainer, mapStateToProps, mapDispatchToProps } from '../MovieCardContainer';
import { setModal, getTrailer, getMovieDetails } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  setModal: jest.fn().mockReturnValue(false),
  getTrailer: jest.fn().mockReturnValue('trailer'),
  getMovieDetails: jest.fn().mockReturnValue('details'),
}))

describe('MovieCardContainer', () => {
  describe('component', () => {
    test('should render without crashing', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<MovieCardContainer />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });

    test('should render other elements, depends on props', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<MovieCardContainer viewMode="gallery"/>);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });
  })

  describe('mapStateToProps', () => {
    test('should return the right value', () => {

      const initialState = {
        viewMode: false,
      };

      expect(mapStateToProps(initialState).viewMode).toEqual(false);
    });
  });

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onTrailerButtonClick should dispatch setModal action', () => {
      const dispatch = jest.fn();
      const { onTrailerButtonClick } = mapDispatchToProps(dispatch);

      onTrailerButtonClick();

      expect(setModal).toHaveBeenCalled();
      expect(getTrailer).toHaveBeenCalled();
      expect(dispatch).toHaveBeenCalled();
    });

    test('onLinkClick should dispatch getMovieDetails action', () => {
      const dispatch = jest.fn();
      const { onLinkClick } = mapDispatchToProps(dispatch);

      onLinkClick();

      expect(getMovieDetails).toHaveBeenCalled();
      expect(dispatch).toHaveBeenCalled();
    })
  })
})
