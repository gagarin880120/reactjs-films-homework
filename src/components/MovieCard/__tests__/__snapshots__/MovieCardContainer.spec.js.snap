// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCardContainer component should render without crashing 1`] = `
<MovieCard
  movie={Object {}}
  onTrailerButtonClick={null}
/>
`;
